src/java.base/share/classes/sun/nio/ch/SelectorImpl.java

```diff
@@ -127,7 +129,16 @@ private int lockAndDoSelect(Consumer<SelectionKey> action, long timeout)
             inSelect = true;
             try {
                 synchronized (publicSelectedKeys) {
-                    return doSelect(action, timeout);
+                    if (!SelectionEvent.enabled()) {
+                        return doSelect(action, timeout);
+                    }
+                    long start = SelectionEvent.timestamp();
+                    int n = doSelect(action, timeout);
+                    long duration = SelectionEvent.timestamp() - start;
+                    if (SelectionEvent.shouldCommit(duration)) {
+                        SelectionEvent.commit(start, duration, n);
+                    }
+                    return n;
```

Comment:
I'd prefer if the existing code moved to implLockAndDoSelect so that lockAndDoSelect commits the event outside of the locked region.